Bottom: 21c4a9a343ca12d8b01ae5beff2ffa49fbbd8a76
Top:    ff6b57b8ce3af09323bea7f9e8ed4c0a1450a339
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-06 21:14:08 +0100

Refresh of give-the-hids-control-over-att

---

diff --git a/src/hid/gtk/gtkhid-gdk.c b/src/hid/gtk/gtkhid-gdk.c
index bf42d86..e72aa65 100644
--- a/src/hid/gtk/gtkhid-gdk.c
+++ b/src/hid/gtk/gtkhid-gdk.c
@@ -780,22 +780,54 @@ ghid_invalidate_all ()
 void
 ghid_notify_crosshair_change (bool changes_complete)
 {
+  static int invalidate_depth = 0;
+
   /* FIXME: We sometimes get called before the GUI is up */
   if (gport->drawing_area == NULL)
     return;
 
   if (changes_complete)
+    invalidate_depth --;
+
+  if (invalidate_depth < 0)
+    {
+      fprintf (stderr, "ERROR: Unmatched notify_crosshair_change calls\n");
+      invalidate_depth = 0;
+    }
+
+  if (invalidate_depth == 0)
+    DrawAttached ();
+
+  if (!changes_complete)
+    invalidate_depth ++;
+  else if (gport->drawing_area != NULL)
     ghid_draw_area_update (gport, NULL);
 }
 
 void
 ghid_notify_mark_change (bool changes_complete)
 {
+  static int invalidate_depth = 0;
+
   /* FIXME: We sometimes get called before the GUI is up */
   if (gport->drawing_area == NULL)
     return;
 
   if (changes_complete)
+    invalidate_depth --;
+
+  if (invalidate_depth < 0)
+    {
+      fprintf (stderr, "ERROR: Unmatched notify_mark_change calls\n");
+      invalidate_depth = 0;
+    }
+
+  if (invalidate_depth == 0)
+    DrawMark ();
+
+  if (!changes_complete)
+    invalidate_depth ++;
+  else
     ghid_draw_area_update (gport, NULL);
 }
 
diff --git a/src/hid/lesstif/main.c b/src/hid/lesstif/main.c
index 5f3da73..2a27086 100644
--- a/src/hid/lesstif/main.c
+++ b/src/hid/lesstif/main.c
@@ -2882,43 +2882,43 @@ lesstif_invalidate_all (void)
 static void
 lesstif_notify_crosshair_change (bool changes_complete)
 {
-  static int invalidate_count = 0;
+  static int invalidate_depth = 0;
 
   if (changes_complete)
-    invalidate_count --;
+    invalidate_depth --;
 
-  if (invalidate_count < 0)
+  if (invalidate_depth < 0)
     {
       fprintf (stderr, "ERROR: Unmatched notify_crosshair_change calls\n");
       invalidate_depth = 0;
     }
 
-  if (invalidate_count == 0)
+  if (invalidate_depth == 0)
     DrawAttached ();
 
   if (!changes_complete)
-    invalidate_count ++;
+    invalidate_depth ++;
 }
 
 static void
 lesstif_notify_mark_change (bool changes_complete)
 {
-  static int invalidate_count = 0;
+  static int invalidate_depth = 0;
 
   if (changes_complete)
-    invalidate_count --;
+    invalidate_depth --;
 
-  if (invalidate_count < 0)
+  if (invalidate_depth < 0)
     {
       fprintf (stderr, "ERROR: Unmatched notify_mark_change calls\n");
       invalidate_depth = 0;
     }
 
-  if (invalidate_count == 0)
+  if (invalidate_depth == 0)
     DrawMark ();
 
   if (!changes_complete)
-    invalidate_count ++;
+    invalidate_depth ++;
 }
 
 static int
